[W-10874727] - Elevate NonElevate Filter

gerardorodriguezsf opened the pull request:
# Technical Design Review
[[Technical Design Review] W-10874727 - NPSP](https://salesforce.quip.com/MweMAiEOgjH1)

# Changes
Updated the RD2_ETableController to get the ‘Recurring Donation’ LWC field ‘Show these types of Recurring Donations’ with the selected option from the Experience Builder. Updated RD2_ERecurringDonationsSelector to return only the filtered recurring donation records.

# Work Closed
[W-10874727](https://gus.lightning.force.com/a07EE00000tLepiYAC)

maurimart reviewed (approved):
LGTM

force2b commented on force-app/main/default/classes/RD2_ERecurringDonationsSelector.cls:
```diff
@@ -35,6 +35,10 @@
 
+	private final String ALL_RD = 'All Recurring Donations';
+	private final String ELEVATE_RD = 'Elevate Recurring Donations';
+	private final String NONELEVATE_RD = 'Non-Elevate Recurring Donations';
```
These look like hardcoded picklist value strings. Where are these defined for use elsewhere?

gerardorodriguezsf replied:
Hello Michael, these values were defined [here](https://salesforce.quip.com/kRYaAjEE8aHr?range_cols=1&range_rows=4#temp:s:temp:C:PQJ39e645cd36ea4f44919fc2850;temp:C:PQJf83d97a2a7bc4183b0315cc97). They are used [here](https://github.com/SalesforceFoundation/NPSP/blob/feature/240__filter_rd_elevate_nonelevate/force-app/main/default/lwc/rd2RecurringDonation/rd2RecurringDonation.js-meta.xml) To show this options on the Experience Builder, and then only show the Recurring Donation records based on this filter.
Please let me know if you need help here.

force2b replied:
Thanks. I see they're related to text in the LWC component config. I was trying to check on whether those can ever be translated, but it doesn't look like that's the case. So, these are probably fine - at least for now.

force2b reviewed (changes requested):
One question on hardcoded strings, but this PR also has a merge conflict that needs resolution in order to complete the review.

JoseRossiSF reviewed (approved):
LGTM

voduyemi commented on force-app/main/default/classes/RD2_ERecurringDonationsSelector.cls:
```diff
@@ -50,27 +54,35 @@ public inherited sharing class RD2_ERecurringDonationsSelector {
      * @return List<npe03__Recurring_Donation__c>
      */
-    public List<npe03__Recurring_Donation__c> getRecurringDonations() {
+    public List<npe03__Recurring_Donation__c> getRecurringDonations(String elevateFilter) {
```
Have we considered using dynamic SOQL here? That way we are not having to loop through all the RDs returned and filter based on `CommitmentId__c`.

gerardorodriguezsf replied:
Hi @voduyemi , thank you very much for your feedback. I updated this method based on your feedback, Can you check it again?

voduyemi commented on force-app/main/default/classes/RD2_ETableController_TEST.cls:
```diff
@@ -111,7 +114,7 @@ public class RD2_ETableController_TEST {
+        List<RD2_ETableController.TableView> result = RD2_ETableController.retrieveTableView(RD2_ETableController_TEST.ALL_RD);
         Test.stopTest();
 
         System.assertNotEquals(null, result, 'Recurring Donations should not be null');
```
Can we assert that the values in the TableView object are set correctly? The process of building this object calls methods that run logic to change the set values, we would benefit greatly from asserts to confirm that `frequency`,  `status` and `paymentMethod` values are set correctly.

gerardorodriguezsf replied:
Yes @voduyemi I double checked it, and I also created one new scratch org with this code implementations to check the complete flow, and check this values too.

force2b commented on force-app/main/default/classes/RD2_ERecurringDonationsSelector.cls:
```diff
@@ -50,29 +53,38 @@ public inherited sharing class RD2_ERecurringDonationsSelector {
-                  					AND RecurringType__c IN (:RD2_Constants.RECURRING_TYPE_OPEN, :RD2_Constants.RECURRING_TYPE_FIXED) 
-									WITH SECURITY_ENFORCED 
-									ORDER BY npe03__Next_Payment_Date__c DESC NULLS LAST];
+			String queryString = 'SELECT npe03__Amount__c, '
```
For indentations in this file - and any file in the repo - please use 4 spaces for indentations and not tabs. Looking at this PR, it's clear the files now have a mix of spaces and tabs for indentation. You should be able to use your IDE to fix the indentations and the recommit. Please ensure that all of your team members are also using spaces and not tabs. Consistency is critical in a product of this size.

gerardorodriguezsf replied:
Thanks Micheal, I'll have this into account. Let me fix this.

force2b reviewed (changes requested):
I'm still seeing tabs everywhere. Please check all of the files in this PR as well as with your other teammates to make sure that nobody else is using tabs

gerardorodriguezsf commented:
> 

Thanks Michael, I fixed the indentation in the worked files from this PR.
I converted indentations to spaces, and set them to use 4 spaces.
I already alerted the team.

force2b reviewed (changes requested):
This has a [build failure](https://metaci.sfdc.sh/builds/644527) due to a class compilation issue. Same as with the 6991 PR.

alvaroolivenciasf commented on force-app/main/default/classes/RD2_ERecurringDonationsSelector.cls:
```diff
@@ -50,32 +53,39 @@ public inherited sharing class RD2_ERecurringDonationsSelector {
+                                + 'npe03__Next_Payment_Date__c '
+                                + 'FROM npe03__Recurring_Donation__c '
+                                + 'WHERE npe03__Contact__c = \'' + currentContactId + '\' '
+                                + 'AND RecurringType__c IN (\''+RD2_Constants.RECURRING_TYPE_OPEN+'\',\''+RD2_Constants.RECURRING_TYPE_FIXED+'\') ';
```
@force2b as a CRM, I can create a RD where the RecurringType__c is equal to a value I define in the picklist, i.g. "random". By doing this we are enforcing that only these 2 are displayed in the table. Does this make sense to you ?

alvaroolivenciasf replied:
@gerardorodriguezsf it was decided that the `AND RecurringType__c IN` should be removed from the query

gerardorodriguezsf replied:
Ok @force2b, thank you very much for the clarification, I removed it and also alerted the team about this.

alvaroolivenciasf replied:
@lparrott @voduyemi  is it possible you continue with Michael’s change request and review @gerardorodriguezsf's reply ? Michael is on PTO during the next week and we would appreciate to finish with this PR during the first days of the next week.

alvaroolivenciasf commented:
@gerardorodriguezsf @voduyemi @gerardorodriguezsf we are getting this of this PR: @force2b told me that admins shouldn't add a custom option into RecurringType__c, if they do this, they break NPSP. Please cancel this PR

lparrott commented on force-app/main/default/classes/RD2_ERecurringDonationsSelector.cls:
```diff
@@ -50,34 +53,40 @@ public inherited sharing class RD2_ERecurringDonationsSelector {
+            }
 
+            queryString += 'WITH SECURITY_ENFORCED ';
+            queryString += 'ORDER BY npe03__Next_Payment_Date__c DESC NULLS LAST';
```
It looks like you're setting the queryString, but not actually calling the query?
